Encryption at rest of manager keys and raft data

aaronlehmann opened the pull request:
This implements the Docker side of https://github.com/docker/swarmkit/pull/1598. Builds on earlier work by @tonistiigi.

It's feature-complete but needs docs and some more testing.

cc @cyli @diogomonica @NathanMcCauley

tonistiigi committed:
vendor: update swarmkit to bddd3f0

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

tonistiigi committed:
vendor: update errors to 839d9e91

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

tonistiigi committed:
Add support for swarm init lock and swarm unlock

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

tonistiigi committed:
Add integration tests for encrypted swarm

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

Commit:
Revise swarm init/update flags, add unlocking capability

- Neither swarm init or swarm update should take an unlock key
- Add an autolock flag to turn on autolock
- Make the necessary docker api changes
- Add SwarmGetUnlockKey API call and use it when turning on autolock
- Add swarm unlock-key subcommand

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Add unlock key rotation

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Lock/unlock integration tests

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

thaJeztah commented:
needs a rebase @aaronlehmann

aaronlehmann commented:
Gah! It's not ready to merge. Super annoying how CI forces a merge against master and then falls over with compilation errors when there are conflicts.

aaronlehmann commented:
ping @dnephin @thaJeztah @aluzzardi - it would be great to get some early design review

aluzzardi commented on api/server/router/swarm/cluster.go:
```diff
@@ -28,7 +28,9 @@ func (sr *swarmRouter) initRoutes() {
 		router.NewPostRoute("/swarm/join", sr.joinCluster),
 		router.NewPostRoute("/swarm/leave", sr.leaveCluster),
 		router.NewGetRoute("/swarm", sr.inspectCluster),
+		router.NewGetRoute("/swarm/unlockkey", sr.getUnlockKey),
```
Is this route supposed to return the unlock key or to rotate it

aaronlehmann replied:
> Is this route supposed to return the unlock key or to rotate it?

Return the unlock key

aluzzardi commented:
What's the overall workflow UX-wise?

What happens while the swarm in locked?

aaronlehmann commented on hack/vendor.sh:
```diff
@@ -1,15 +0,0 @@
-#!/bin/bash
```
Fixing

aaronlehmann commented:
> What's the overall workflow UX-wise?

`docker swarm init --autolock` or `docker swarm update --autolock` to turn on manager locking. This prints out a key.

Then on restart `docker swarm unlock` is necessary to start the manager. This takes the key on stdin.

You can retrieve the current key with `docker swarm unlock-key`, or rotate it with `docker swarm unlock-key --rotate`.

> What happens while the swarm in locked?

The swarm components do not operate until you run `docker swarm unlock`.

aluzzardi commented:
This needs help reviewing, specifically the UX workflow (see @aaronlehmann comment above).

Some comments:

- `autolock` seems weird - there is no `lock` command, so it's not obvious what we're automating. `autounlock` would maybe make more sense? It's symmetric with `docker swarm unlock`. /cc @dnephin 
- Exposing the CA service and having the Engine talk to it seems weird. We should not talk to anything else but the Control API IMHO. Perhaps we should just return the unlock key in `inspect` and remove the unlockkey endpoint altogether?
- Consistency: We redact sensitive data from the API (such as secret data), yet we allow retrieving the unlockkey. Does this constitute a security issue? /cc @diogomonica 

/cc @icecrime @thaJeztah @vieux @tiborvass

tonistiigi commented:
Tested and reviewed this and design LGTM. Previous version had a way to specify lock key and never returned it from remote api. I think it was changed for ease of use and I don't really have a preference about which is better.

The behavior of the down nodes after rotation/autolock change needs to be clearly documented as it isn't obvious.

@aaronlehmann Maybe in follow-up we could run all swarm tests with encryption enabled. Even if it requires an env switch it would give much more coverage.

aluzzardi commented:
@icecrime @thaJeztah @vieux @tiborvass @diogomonica: Any thoughts regarding the comments above?

This is getting merged today, feedback much appreciated.

vieux commented:
@aluzzardi `--autolock` seems ok to me, I don't have a better name

@aaronlehmann could you add API and CLI reference doc  ?

vieux commented:
@aaronlehmann needs rebase too

Commit:
Document autolock/unlock/unlock-key

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "encryption" git@github.com:aaronlehmann/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842354230856
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

aaronlehmann commented:
@vieux: Thanks. I've rebased (squashed into fewer commits) and added API/CLI reference docs. Also, the swarmkit PR has been merged upstream, so this should be ready to go.

I'm happy to make more changes in followup PRs after the merge.

PTAL

aluzzardi commented:
LGTM

cyli commented:
LGTM

vieux commented:
LGTM

thaJeztah commented on docs/reference/api/docker_remote_api_v1.25.md:
```diff
@@ -4762,6 +4765,9 @@ Initialize a new swarm. The body of the HTTP response includes the node ID.
         "CAConfig": {}
+        "EncryptionConfig" : {
+          "AutoLockManagers": false
+        },
```
for a follow up; remove the trailing comma, and add a comma after `"CAConfig": {}`

aaronlehmann replied:
@thaJeztah: done

thaJeztah commented on docs/reference/commandline/swarm_update.md:
```diff
@@ -21,6 +21,7 @@ Usage:  docker swarm update [OPTIONS]
 Update the swarm
 
 Options:
+      --autolock                        Enable or disable manager autolocking (requiring an unlock key to start a stopped manager)
```
for follow-up: add `bool` here

aaronlehmann replied:
@thaJeztah: `bool` doesn't appear in the `--help` output. Should I add it anyway.

thaJeztah replied:
Just realized that, hm probably in the comment I guess

thaJeztah commented on docs/reference/commandline/swarm_init.md:
```diff
@@ -22,6 +22,7 @@ Initialize a swarm
 
 Options:
       --advertise-addr value            Advertised address (format: <ip|interface>[:port])
+      --autolock                        Enable or disable manager autolocking (requiring an unlock key to start a stopped manager)
```
same here (`bool`)

thaJeztah replied:
also needs to mention the default if omitted

aaronlehmann replied:
@thaJeztah: Can you suggest a phrasing for this? `--autolock` or `--autolock=true` will enable it, and `--autolock=false` will disable it (in cases where it was previously enabled).

thaJeztah replied:
This one is for `init`, so it can only "enable" it, so this can be simply;

```
Enable manager autolocking (requiring an unlock key to start a stopped manager)
```

aaronlehmann commented:
>  Options:
>        --advertise-addr value            Advertised address (format: <ip|interface>[:port])
> +      --autolock                        Enable or disable manager autolocking (requiring an unlock key to start a stopped manager)
> This one is for init, so it can only "enable" it, so this can be simply;
> 
> Enable manager autolocking (requiring an unlock key to start a stopped manager)

The flag is shared between init and update right now. Do you think the phrasing of the description should be different for these connands?

aaronlehmann commented:
Saw this failure in CI running on an unrelated PR:

```
02:44:57 ----------------------------------------------------------------------
02:44:57 FAIL: docker_cli_swarm_test.go:952: DockerSwarmSuite.TestSwarmRotateUnlockKey
02:44:57 
02:44:57 [d761dbbbf650b] waiting for daemon to start
02:44:57 [d761dbbbf650b] daemon started
02:44:57 [d761dbbbf650b] exiting daemon
02:44:57 [d761dbbbf650b] waiting for daemon to start
02:44:57 [d761dbbbf650b] daemon started
02:44:57 [d761dbbbf650b] exiting daemon
02:44:57 [d761dbbbf650b] waiting for daemon to start
02:44:57 [d761dbbbf650b] daemon started
02:44:57 docker_cli_swarm_test.go:1020:
02:44:57     c.Assert(err, checker.IsNil, check.Commentf("out: %v", string(out)))
02:44:57 ... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc421b583c0), Stderr:[]uint8(nil)} ("exit status 1")
02:44:57 ... out: Error response from daemon: swarm could not be unlocked: invalid key provided
02:44:57 
02:44:57 
02:44:57 [d761dbbbf650b] exiting daemon
```

I'll investigate this and provide a fix.
